### ISSUE TITLE
If an authorization code is exchanged multiple times, revoke all access tokens already granted based on the authorization code

### ISSUE THREAD

**Issue by shane-tomlinson** [mozilla/fxa#987]
In the [Authorization Codes section](https://tools.ietf.org/html/rfc6749#section-10.5) of the OAuth spec, they have this to say about multiple exchanges for the same authorization code:

> If the authorization server observes multiple attempts to exchange an
   authorization code for an access token, the authorization server
   SHOULD attempt to revoke all access tokens already granted based on
   the compromised authorization code.

I don't think we can actually do this. We [delete the Authorization code as soon it's checked](https://github.com/mozilla/fxa/blob/3fd7d05191185bb7c5d5f6a9b8a31b336826a96c/packages/fxa-auth-server/fxa-oauth-server/lib/routes/token.js#L324) and we don't store the authorization code's value with [the access token](https://github.com/mozilla/fxa/blob/3fd7d05191185bb7c5d5f6a9b8a31b336826a96c/packages/fxa-auth-server/fxa-oauth-server/lib/db/mysql/index.js#L445). So likely unless we start storing more data, there's nothing we can do here and should just close this down.

@rfk - thoughts?

**Comment by rfk** [comment]
I agree with your assessment, we can't do it with the current db schema.

We could consider adding a "used" boolean column to the codes table instead of just deleting them, which would allow us to implement this effectively. Code are short-lived so we could prune them after e.g. an hour and still get some benefit from this approach without the db growing too large.  (In fact, do we have pruning for expired codes in place right now?)

**Comment by shane-tomlinson** [comment]
>  (In fact, do we have pruning for expired codes in place right now?)

@jrgm - are we pruning expired codes?

**Comment by shane-tomlinson** [comment]
> > (In fact, do we have pruning for expired codes in place right now?)

> @jrgm - are we pruning expired codes?

Answer in a meeting, no, we are not pruning expired codes. 

Opened #1044 to prune expired codes.
